pennylane/shadows/classical_shadow.py

```diff
@@ -0,0 +1,370 @@
+            word = pauli_list_to_word([observable])
+            return [(1, word)]
+
+        if isinstance(observable, qml.operation.Tensor):
+            word = pauli_list_to_word(observable.obs)
+            return [(1, word)]
+
+        # TODO: cases for new operator arithmetic
+
+        if isinstance(observable, qml.Hamiltonian):
+            coeffs_and_words = []
+            for coeff, op in zip(observable.data, observable.ops):
+                coeffs_and_words.extend(
+                    [(coeff * c, w) for c, w in self._convert_to_pauli_words(op)]
+                )
+            return coeffs_and_words
+
+    def expval(self, H, k):
+        r"""Compute expectation value of an observable :math:`H`.
+
+        The canonical way of computing expectation values is to simply average the expectation values for each local snapshot, :math:`\langle O \rangle = \sum_t \text{tr}(\rho^{(t)}O) / T`.
+        This corresponds to the case ``k=1``. However, it is often desirable for better accuracy to split the ``T`` measurements into ``k`` equal parts to compute the median of means, see `2002.08953 <https://arxiv.org/abs/2002.08953>`_.
+
+        One of the main perks of classical shadows is being able to compute many different expectation values by classically post-processing the same measurements. This is helpful in general as it may help
+        save quantum circuit executions.
+
+        Args:
+            H (:class:`~.pennylane.Hamiltonian` or :class:`~.pennylane.operation.Tensor`): Observable to compute the expectation value over.
+            k (int): Number of equal parts to split the shadow's measurements to compute the median of means. ``k=1`` corresponds to simply taking the mean over all measurements.
+
+        Args:
+            H (qml.Observable): Observable to compute the expectation value
+            k (int): Split the snapshots into ``k`` equal parts to compute the median of means.
+
+        Returns:
+            float: expectation value estimate.
+
+        **Example**
+
+        .. code-block:: python3
+
+            dev = qml.device("default.qubit", wires=range(2), shots=1000)
+            @qml.qnode(dev)
+            def qnode(x):
+                qml.Hadamard(0)
+                qml.CNOT((0,1))
+                qml.RX(x, wires=0)
+                return classical_shadow(wires=range(2))
+
+            bits, recipes = qnode(0)
+            shadow = ClassicalShadow(bits, recipes)
+
+        Compute Pauli string observables
+
+        >>> shadow.expval(qml.PauliX(0) @ qml.PauliX(1), k=1)
+        (1.0079999999999998+0j)
+
+        or of a Hamiltonian using `the same` measurement results
+
+        >>> H = qml.Hamiltonian([1., 1.], [qml.PauliZ(0)@qml.PauliZ(1), qml.PauliX(0)@qml.PauliX(1)])
+        >>> shadow.expval(H, k=1)
+        (2.2319999999999998+0j)
+        """
+        coeffs_and_words = self._convert_to_pauli_words(H)
+
+        expval = 0
+        for coeff, word in coeffs_and_words:
```

Comment:
I've implemented this vectorization in the state differentiability PR, since that's the one with the most obvious benefits (exponentially many observables!).

Do you want me to move that implementation to this PR?

Reply:
Would it be a hassle? Having the option to return multiple observables already on the class level seems nice to have as well if its not too much trouble!

Reply:
Added 👍 .